cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,950 @@
+
+  // The maximum floating-point fold supported by the library
+  static constexpr auto MAXFOLD = MAXINDEX + 1;
+
+private:
+  /// Binned floating-point compression factor
+  /// This factor is used to scale down inputs before deposition into the bin of
+  /// highest index
+  static constexpr auto COMPRESSION = 1.0 / (1 << (MANT_DIG - BIN_WIDTH + 1));
+  /// Binned double precision expansion factor
+  /// This factor is used to scale up inputs after deposition into the bin of
+  /// highest index
+  static constexpr auto EXPANSION = 1.0 * (1 << (MANT_DIG - BIN_WIDTH + 1));
+  static constexpr auto EXP_BIAS  = MAX_EXP - 2;
+  static constexpr auto EPSILON   = ::cuda::std::numeric_limits<ftype>::epsilon();
+  /// Binned floating-point deposit endurance
+  /// The number of deposits that can be performed before a renorm is necessary.
+  /// Applies also to binned complex double precision.
+  static constexpr auto ENDURANCE = 1 << (MANT_DIG - BIN_WIDTH - 2);
+
+  /// Return a binned floating-point bin
+  _CCCL_DEVICE _CCCL_FORCEINLINE ftype binned_bins(int index) const
+  {
+    ftype* bins = get_shared_bin_array<ftype, MAXINDEX + MAXFOLD>();
+    return bins[index];
+  }
+
+  /// Get the bit representation of a float
+  _CCCL_DEVICE _CCCL_FORCEINLINE static uint32_t& get_bits(float& x)
+  {
+    return *reinterpret_cast<uint32_t*>(&x);
+  }
+  /// Get the bit representation of a double
+  _CCCL_DEVICE _CCCL_FORCEINLINE static uint64_t& get_bits(double& x)
+  {
+    return *reinterpret_cast<uint64_t*>(&x);
+  }
+  /// Get the bit representation of a const float
+  _CCCL_DEVICE _CCCL_FORCEINLINE static uint32_t get_bits(const float& x)
+  {
+    return ::cuda::std::bit_cast<uint32_t>(x);
+  }
+  /// Get the bit representation of a const double
+  _CCCL_DEVICE _CCCL_FORCEINLINE static uint64_t get_bits(const double& x)
+  {
+    return ::cuda::std::bit_cast<uint64_t>(x);
+  }
+
+  /// Return primary vector value const ref
+  _CCCL_DEVICE _CCCL_FORCEINLINE const ftype& primary(int i) const
+  {
+    if (FOLD <= _CUB_RFA_MAX_JUMP_)
+    {
+      switch (i)
+      {
+        case 0:
+          if (FOLD >= 1)
+          {
+            return data[0];
+          }
+          [[fallthrough]];
+        case 1:
+          if (FOLD >= 2)
+          {
+            return data[1];
+          }
+          [[fallthrough]];
+        case 2:
+          if (FOLD >= 3)
+          {
+            return data[2];
+          }
+          [[fallthrough]];
+        case 3:
+          if (FOLD >= 4)
+          {
+            return data[3];
+          }
+          [[fallthrough]];
+        case 4:
+          if (FOLD >= 5)
+          {
+            return data[4];
+          }
+          [[fallthrough]];
+        default:
+          return data[FOLD - 1];
+      }
+    }
+    else
+    {
+      return data[i];
+    }
+  }
+
+  /// Return carry vector value const ref
+  _CCCL_DEVICE _CCCL_FORCEINLINE const ftype& carry(int i) const
```

Comment:
I believe @ericniebler introduced `_CCCL_DEVICE_API` recently, which could replace `_CCCL_DEVICE _CCCL_FORCEINLINE` throughout the PR